pkg/ruler/compat.go

```diff
@@ -78,3 +82,32 @@ func (t *tsdb) StartTime() (int64, error) {
 func (t *tsdb) Close() error {
 	return nil
 }
+
+// engineQueryFunc returns a new query function that executes instant queries against
+// the given engine, after subtracting the provided delay from the instant query timestamp.
+// It converts scalar into vector results.
+// Based on https://github.com/prometheus/prometheus/blob/ecda6013edf58bf645c6661b9f78ccce03b1f315/rules/manager.go#L162-L187
+func engineQueryFunc(engine *promql.Engine, q storage.Queryable, delay time.Duration) rules.QueryFunc {
```

Comment:
This whole function is a copy of `promRules.EngineQueryFunc()` just to add `t = t.Add(-delay)` as first thing. Why don't we simply wrap `promRules.EngineQueryFunc()` instead of duplicating it? Something like (pseudo-code):

```
func engineQueryFunc(engine *promql.Engine, q storage.Queryable, delay time.Duration) rules.QueryFunc {
orig := promRules.EngineQueryFunc(engine, queryable)

return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
  return orig(ctx, qs, t.Add(-delay))
}
}
```

Reply:
+1

Reply:
Will do.